discord/voice_client.py

```diff
@@ -643,3 +691,163 @@ def send_audio_packet(self, data, *, encode=True):
+        t.start()
+
+    def stop_recording(self):
+        """Stops the recording.
+
+        Must be already recording.
+
+        Raises
+        ------
+        ClientException
+            Not currently recording.
+        """
+        if not self.recording:
+            raise ClientException("Not currently recording audio.")
+        self.decoder.stop()
+        self.recording = False
+        self.paused = False
+
+    def pause_recording(self):
+        """Pauses or unpauses the recording.
+
+        Must be already recording.
+
+        Raises
+        ------
+        ClientException
+            Not currently recording.
+         """
+        if not self.recording:
+            raise ClientException("Not currently recording audio.")
+        self.paused = {True: False, False: True}[self.paused]
+
+    def empty_socket(self):
+        while True:
+            ready, _, _ = select.select([self.socket], [], [], 0.0)
+            if len(ready) == 0:
+                break
+            for s in ready:
+                s.recv(4096)
+
+    def recv_audio(self, sink, callback, *args):
+        #  Gets data from _recv_audio and sorts
+        #  it by user, handles pcm files and
+        #  silence that should be added.
+
+        self.user_timestamps = {}
+        self.starting_time = time.perf_counter()
+        while self.recording:
+            ready, _, err = select.select([self.socket], [],
+                                          [self.socket], 0.01)
+            if not ready:
+                if err:
+                    print("Socket error")
+                continue
+
+            try:
+                data = self.socket.recv(4096)
+            except OSError:
+                self.stop_recording()
+                continue
+
+
+            self.unpack_audio(data)
+
+        self.stopping_time = time.perf_counter()
+        self.sink.cleanup()
+        try:
+            callback = asyncio.run_coroutine_threadsafe(callback(self.sink, *args), self.loop)
+            result = callback.result()
+        except Exception as exc:
+            raise exc
+        else:
+            if result is not None:
+                print(result)
+
+    def ssrc_exists(self, ssrc):
+        return ssrc in self.ws.ssrc_map
+
+    def recv_decoded_audio(self, data):
+        if data.ssrc not in self.user_timestamps:
+            self.user_timestamps.update({data.ssrc: data.timestamp})
+            # Add silence of when they were not being recorded.
+            data.decoded_data = struct.pack('<h', 0) * round(self.decoder.CHANNELS * self.decoder.SAMPLING_RATE * (time.perf_counter() - self.starting_time)) + data.decoded_data
+        else:
+            self.user_timestamps[data.ssrc] = data.timestamp
+
+        silence = data.timestamp - self.user_timestamps[data.ssrc] - 960
+        data.decoded_data = struct.pack('<h', 0) * silence + data.decoded_data
+        while not self.ssrc_exists(data.ssrc):
+            time.sleep(0.2)
```

Comment:
Yes though I wouldn't want to run even more threads. The ssrc usually exists by the time the data is decoded, if not then it usually only takes less than a second for it to be received by Discord so it shouldn't block the event loop very much. If it does become a problem later then maybe it can be changed but at least for now it's fine.

Reply:
Do keep in mind that larger bots have hundreds of concurrent async operations. Blocking the event loop for a second can be catastrophic to other functions. 
If you don’t want another thread, async is your friend. And there’s nothing wrong with using async, since the library requires it anyways.

Reply:
The only thread that it would be holding up is the decoding thread which is in charge of decoding data in the background. The bot would still run perfectly fine for everything else including receiving and decrypting the audio packets from Discord. I also can't use asyncio since everything involving recording is non-asynchronous.

Reply:
Are you familiar with run_in_executor or run_until_complete?

Reply:
ah right, I forgot about that. I don't use asyncio very often.